### ISSUE TITLE
Allow login using username as well

### ISSUE THREAD

**Issue by MrPetovan** [pixelfed/pixelfed#494]
# Expected behavior

I can log in with my email address or my username.

# Actual behavior

I can only use my email address to log in.

# Rationale

Usernames are unique as well as email addresses, which makes them a valid identifier to log in. And they usually are shorter, which makes them more convenient.

**Comment by trwnh** [comment]
Something to consider: usernames are public, whereas emails are private. Being able to login with username is inherently less secure, as an attacker now no longer needs to know which email you used for signup.

**Comment by MrPetovan** [comment]
How are emails private?

**Comment by trwnh** [comment]
Emails are private because you can't tell which one is linked to a given account. If I wanted to hack pixelfed.social/dansup, then I could theoretically guess passwords until one of them works. Or alternatively, maybe I find out dansup's password somehow. Then, I can log into dansup's account using the password I obtained + the username which is public knowledge. But if you can only log in with an email, then I now have to figure out which email was used for that account as well. The email is not public knowledge, it is not listed on the profile, etc.

**Comment by MrPetovan** [comment]
I don't think an email address is inherently private in a security sense because, unlike a password, you have to give it out to other people to communicate with you. You can personally coerce emails into being private by not using email with other people and by using a unique address by service provider, but this isn't what an email address is about. It is a private piece of data, but it isn't a secure piece of data.

Allowing only email addresses as identifier is security by obscurity, which is only marginally better than no security, and probably not worth the lack of convenience. If you want real security, 2FA is the way, not wrongly assuming email addresses are private.

Besides, right now I can hit up the password reset form to test for the use of any email address on a given Pixelfed server, which kind of busts the "privacy" of email addresses.

Bottom line: there are ways of really securing Pixelfed accounts, and none of them include disallowing username as identifier.

**Comment by trwnh** [comment]
You are correct that it is not fully "secure", but incorrect that it is not "private". A properly designed password-reset form would give no indication that you've stumbled upon a valid email -- it would instead respond the same to any email address: "If this is a valid email address with an account linked to it, we have sent the reset details to that email." This way, an attacker does not know whether the reset attempt was successful.

In fact, going one step further, we can see that passwords themselves are not secure, either! What is a password, if not "security by obscurity", as you put it? However, the need to know both the email AND the password can be considered a sort of two- ~factor~ element authentication by default, since you must input two of "something I know". If logging in with a username were to be supported, then it might be a good idea to request (or enforce) that the user enable a second authentication factor to replace the use of email as an authentication factor.

**Comment by MrPetovan** [comment]
An email address. Isn't. An authentication. Factor. 

It isn't "something I know".  It is "something I and everyone I've ever sent or received an email to/from know". It *may* be "something only I know", but it wasn't designed that way (unlike passwords) and for that reason shouldn't be considered that way either.

And even if it was, 2FA scenarios never add two "something I know".

Please reconsider your view on email addresses being secret or secure, they aren't by design and therefore shouldn't be treated as such, and even if they were the combination email+password still wouldn't be 2FA.

Agreed on hardening the password reset form and suggesting 2FA, but it has absolutely nothing to do with enabling usernames as a login identifier which isn't supposed nor required to be secret.

**Comment by dansup** [comment]
@MrPetovan Username authentication is not supported out of the box by the laravel framework we are using. Will consider adding this.

**Comment by MrPetovan** [comment]
I found this if it can help hasten this task: ~~https://laracasts.com/discuss/channels/general-discussion/log-in-with-username-or-email-in-laravel-5~~

**Comment by dansup** [comment]
@MrPetovan That answer is outdated based on the version we are using, ~I will make this an optional feature (default to email only) and push it in the next commits!~

**Comment by MrPetovan** [comment]
Thanks, I started to look into it myself, but I'm completely new to Laravel.

**Comment by trwnh** [comment]
I just want to clarify a few things that I may not have made clear enough:

>An email address. Isn't. An authentication. Factor. It isn't "something I know". It is "something I and everyone I've ever sent or received an email to/from know". It may be "something only I know", but it wasn't designed that way (unlike passwords) and for that reason shouldn't be considered that way either.

Email addresses, usernames, and passwords are all authentication factors, specifically knowledge factors. But in the context of a pixelfed profile, **the username is _public_ knowledge.** Email addresses are not exposed anywhere in the profile, and you have no way of knowing which email is linked to a given account.

If you claim this is untrue, then please tell me which email is linked to pixelfed.social/trwnh -- you may be able to guess based on your knowledge of which emails I have ever used, but you do not have 100% certainty in the same way that you know for sure that my username is `trwnh`. You will be forced to try multiple emails that I have made public, and you may still find that I am using an email that I have not publicized.

> And even if it was, 2FA scenarios never add two "something I know".

I plead ambiguity :) You are correct that a proper setup should use more than one type of factor; what I meant to say was that **having to use an email and a password is two _element_ authentication, within a single _factor_ (knowledge)**. Whereas being able to use a username and a password is more like single-element authentication, because usernames are publicized in profiles.

> Please reconsider your view on email addresses being secret or secure, they aren't by design and therefore shouldn't be treated as such

Email addresses are still more secret than usernames, as I've hopefully made clear. I have not claimed them to be secure, however -- an easily-guessable email is about as secure as an easily-guessable password. **But even "easily-guessed" is marginally more secure than "public information".**

This is why I cannot advocate for allowing username login _by default_, as it represents a net loss of security for the overall network. Defaults matter. I don't think it's wise to allow users slightly more convenience at the cost of having only one knowledge factor guarding their account, which is why I think that if this is implemented, it should (1) be optional, and (2) _heavily_ encourage the use of another factor to make up for this reduction. At minimum, it should warn the user (and/or admin, depending on the implementation) that they are placing accounts at marginally more risk by allowing username login.

**Comment by dansup** [comment]
@MrPetovan @trwnh I have decided against implementing this, and will be fixing the forgot password logic so it doesn't leak invalid emails. It should say "If this email address is associated with an account, a password reset email will be sent" or something.

**Comment by MrPetovan** [comment]
Now I'm curious what about your decision's motivations.

**Comment by dansup** [comment]
@MrPetovan As you pointed out, security. We do not want any data leakage 😉

**Comment by MrPetovan** [comment]
For password recovery, sure, but for login?

**Comment by nicobo** [comment]
I would like to understand the reason too...

My e-mail will probably change in a few years (it did several times already), my pixelfed login don't have to.
Also what about using disposable or unique emails for spam reasons ?

I was surprised that I couldn't 
log-in with my login name at first, which I consider the main identifier of my account...

As @MrPetovan said, 2FA is the way for more security. Third-party authentication mechanisms like OpenID can also provide a way not to reveal the login in the authentication sequence (if you consider it an internal and sensible data), but as it's not implemented yet, using the login sounds logical to me.

Could you reconsider your decision?

**Comment by dansup** [comment]
Your username is public, your email is (usually) not. Allowing username login would make it much easier to bruteforce passwords.

**Comment by roke-julian-lockhart** [comment]
> https://github.com/pixelfed/pixelfed/issues/494#issuecomment-1879776295

@dansup, you've ignored https://github.com/pixelfed/pixelfed/issues/494#issuecomment-621795506 - I use a different https://addy.io/faq/#what-is-a-standard-alias for each site I register for, as is aforedescribed. Consequently, the e-mail address is quite a transient property. Additionally, because my password is automatically generated and 128 characters in length, it cannot be brute-forced (and TOTP 2FA renders that nullified anyway).

**Comment by dansup** [comment]
You can now recover email addresses by account usernames in #4874
